sys/netinet/in_mcast.c

```diff
@@ -485,7 +485,7 @@ imo_multi_filter(const struct ip_moptions *imo, const struct ifnet *ifp,
 	ims = imo_match_source(imf, src);
 
 	if ((ims == NULL && mode == MCAST_INCLUDE) ||
-	    (ims != NULL && ims->imsl_st[0] != mode))
+		(ims != NULL && ims->imsl_st[0] == MCAST_EXCLUDE ))
```

Comment:
I'll fix the style issues here on pushing...
I think this is semantically correct, but I'm not 100% sure since I don't know the code. It matches the comment though and fixes the test cases...

Reply:
Looks right.
If there is no source filter entry => block if that's SSM ("exclude" mode per RFC 3678 clause 3).
If there is an entry => check its action & block if the action is "exclude". 

Though, I'd really really really love to have the tests for all such conditions, similar to the ones provided by the submitter. Multicast code is far from being obvious and having the tests would really help.